plugins/curation/server/__init__.py

```diff
@@ -0,0 +1,198 @@
+        enabled = curation.get(ENABLED, False)
+
+        # update status
+        oldStatus = curation.get(STATUS)
+        if STATUS in params:
+            # regular users can only do construction -> requested transition
+            if curation.get(STATUS) != CONSTRUCTION or \
+               params[STATUS] != REQUESTED:
+                self.requireAdmin(user)
+            curation[STATUS] = params[STATUS]
+        status = curation.get(STATUS)
+
+        # admin enabling curation
+        if enabled and not oldEnabled:
+            # TODO: check if writers exist?
+            # TODO: email writers?
+            folder['public'] = False
+            curation[ENABLE_USER_ID] = user.get('_id')
+            self._addTimeline(oldCuration, curation, 'enabled curation')
+
+        # admin reopening folder
+        if enabled and oldStatus == APPROVED and status == CONSTRUCTION:
+            folder['public'] = False
+            curation[ENABLE_USER_ID] = user.get('_id')
+            self._makeWriteable(folder)
+            self._addTimeline(oldCuration, curation, 'reopened folder')
+
+        # admin disabling curation
+        if not enabled and oldEnabled:
+            self._addTimeline(oldCuration, curation, 'disabled curation')
+
+        # user requesting approval
+        if enabled and oldStatus == CONSTRUCTION and status == REQUESTED:
+            curation[REQUEST_USER_ID] = user.get('_id')
+            self._makeReadOnly(folder)
+            self._addTimeline(oldCuration, curation, 'requested approval')
+            # send email to admin requesting approval
+            self._sendMail(
+                folder, curation.get(ENABLE_USER_ID),
+                'REQUEST FOR APPROVAL: ' + folder['name'],
+                'curation.requested.mako')
+
+        # admin approving request
+        if enabled and oldStatus == REQUESTED and status == APPROVED:
+            folder['public'] = True
+            curation[REVIEW_USER_ID] = user.get('_id')
+            self._addTimeline(oldCuration, curation, 'approved request')
+            # send approval notification to requestor
+            self._sendMail(
+                folder, curation.get(REQUEST_USER_ID),
+                'APPROVED: ' + folder['name'],
+                'curation.approved.mako')
+
+        # admin rejecting request
+        if enabled and oldStatus == REQUESTED and status == CONSTRUCTION:
+            curation[REVIEW_USER_ID] = user.get('_id')
+            self._makeWriteable(folder)
+            self._addTimeline(oldCuration, curation, 'rejected request')
+            # send rejection notification to requestor
+            self._sendMail(
+                folder, curation.get(REQUEST_USER_ID),
+                'REJECTED: ' + folder['name'],
+                'curation.rejected.mako')
+
+        self.model('folder').save(folder)
+        return curation
+
+    def _makeReadOnly(self, folder):
+        for doc in folder['access']['users'] + folder['access']['groups']:
```

Comment:
Here's the API [endpoint](https://github.com/girder/girder/blob/b319a1f9aed8fdd866527e4b2dc3fc5a48220e10/girder/api/v1/folder.py#L219-L240) for updating folder access, you can follow what it does for recursive access updates.  It's a good idea to use a ProgressContext for notifications because these trees could get large.